ipykernel/iostream.py

```diff
@@ -105,87 +109,183 @@ def _check_mp_mode(self):
             return MASTER
         else:
             if not self._have_pipe_out():
-                self._flush_buffer()
                 # setup a new out pipe
                 self._setup_pipe_out()
             return CHILD
 
-    def set_parent(self, parent):
-        self.parent_header = extract_header(parent)
-
+    
+    def start(self):
+        """Start the IOPub thread"""
+        self.thread.start()
+    
+    def stop(self):
+        """Stop the IOPub thread"""
+        self.io_loop.add_callback(self.io_loop.stop)
+        self.thread.join()
+    
     def close(self):
-        self.pub_socket = None
+        self.socket.close()
+        self.socket = None
 
     @property
     def closed(self):
-        return self.pub_socket is None
-
-    def _flush_from_subprocesses(self):
-        """flush possible pub data from subprocesses into my buffer"""
-        if not self._pipe_flag or not self._is_master_process():
-            return
-        for i in range(self._subprocess_flush_limit):
-            if self._pipe_poller.poll(0):
-                msg = self._pipe_in.recv_multipart()
-                if msg[0] != self._pipe_uuid:
-                    continue
-                else:
-                    self._buffer.write(msg[1].decode(self.encoding, 'replace'))
-                    # this always means a flush,
-                    # so reset our timer
-                    self._start = 0
-            else:
-                break
+        return self.socket is None
+    
+    def send_multipart(self, *args, **kwargs):
+        """send_multipart schedules actual zmq send in my thread.
+        
+        If my thread isn't running (e.g. forked process), send immediately.
+        """
 
-    def _schedule_flush(self):
-        """schedule a flush in the main thread
+        if self.thread.is_alive():
+            self.io_loop.add_callback(lambda : self._really_send(*args, **kwargs))
+        else:
+            self._really_send(*args, **kwargs)
+    
+    def _really_send(self, msg, *args, **kwargs):
+        """The callback that actually sends messages"""
```

Comment:
Should the docstring note that there are two different destinations that it might actually send messages to?